designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,209 @@
+---
+title: controller-observation-plugin
+authors:
+  - "@Kavinjsir"
+reviewers:
+  - "@varsha"
+  - "@camila"
+approvers:
+creation-date: 2022-06-06
+last-updated: 2022-06-06
+status: provisional
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# a-grafana-dashboard-plugin-to-observe-controller-metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+- TBD
+
+## Summary
+
+This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.
+
+The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.
+
+In the context of this EP, the Grafana 'manifests' are mainly:
+
+- JSON files to be directly loaded on Grafana dashboards
+- TBD
+
+## Motivation
+
+The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.
+
+Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
+It would be great if we can provide a unified and comprehensive approach so that users may:
+
+- easily visualize controller/operator metrics
+- avoid duplicate creation of Grafana dashboards file
+- improve the observability towards the opeartor, keep it safe and robust
+
+### Goals
+
+- Provide manifests that can display operator status for common needs.
+- Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.
+
+### Non-Goals
+
+- TBD
+
+## Proposal
+
+This proposed implementation is to create a new plugin, `grafana`, that scaffolds the manifests which are loaded by Grafana to display dashboards for controllers' status.
+
+Basically, the manifest can be raw json file such as `controller-dashboard.json` that can be directly applied to Grafana Web UI.
+
+As there are many useful metrics, it might be better to observe in multiple dashboards rather than one.
+
+Hence, we may also provide optional dashboards that focus on different aspects.
+
+For instance: work queue, reconciliation, performance, webhook.
+
+As a result, the layout given by the plugin can be a folder of `config/grafana/`:
+
+```shell
+config/grafana
+├── controller-dashboard.json # The main dashboard raw json file for the simplest usage(copy/paste)
+├── options
+│  ├── performance.json       # CPU/Memory usaage, restful api performance
+│  ├── reconcilier.json       # Reconciliation instruments
+│  ├── webhook.jon            # Webhook Ops
+│  └── workqueue.json         # Work queue status
+└── README.md                 # Docs for the usage
```

Comment:
The specific `.json` file to refer to each metric, right? What happens when that metric is unavailable for a controller. Do we just get a blank dashbaord with no metrics? Also would we be hardcoding the all the metrics that controller-runtime exports? If so, we should probably start by scaffolding only a subset which are really important/widely used.

Reply:
Yeah, @varshaprasad96 's idea is mainly what I thought. Since the controller-runtime provides plenty metrics, it seems not necessary/good to display all of them in one dashboard.

In brief when the queries failed (controller down/authentication failure/syntax error), the panels will be blank. By that time, the dashboard itself is added into the Grafana service, the user may get some error notification when open the page.

I am planning to hardcoding. Since we hope to have the dashboard available on the Grafana market, that means a stable json file, which may be likely as a statice file maintained in the plugin implementation folder?

How would you think of this? Would you have some suggestions or ideas to deal with this more elegantly?